### PR TITLE
Explicitly throw if the array rest/spread items are not iterable.

### LINKED ISSUE #2160: No runtime error for spreading non-iterables
`console.log([...2]);` produces code which outputs an empty array, instead of throwing a runtime error. (Whether or not this is capable of being fixed is unknown to me.)

### PR DESCRIPTION
**Pull request by loganfsmyth**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #2160
| Patch: Bug Fix?          | Y
| Major: Breaking Change?  | Y
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR         | 
| Any Dependency Changes?  |
| License                  | MIT

`Array.from` will just silently create an empty array for non-iterable, non-arraylike objects, so we want to make sure we're actually show an error if a bad this is iterated.

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
What about `arguments`? In old browsers it doesn't have `Symbol.iterator`.

**Comment by Kovensky** [comment]
That's a though one. Probably the only thing we can do is detect the `arguments` pseudo-keyword itself and replace it with `Array.from(arguments)`, because there's nothing else detectable about `arguments`.

**Comment by nicolo-ribaudo** [comment]
Well, we can stringify it (I can't test right know, but if I recall correctly `Object.prototype.toString.call(arguments)` is `[object Arguments]`)

**Comment by loganfsmyth** [comment]
I think I'm up for adding a case to handle arguments. Will give it a shot in a bit.

**Comment by loganfsmyth** [comment]
Alright, this PR got bigger. Might be easiest to follow if you look at the commits individually.

File: packages/babel-traverse/src/scope/index.js [hunk 1]
```diff
@@ -431,14 +431,20 @@ export default class Scope {
       );
     }
 
-    let helperName = "toArray";
+    let helperName;
```
Comment (Andarist) [inline packages/babel-traverse/src/scope/index.js]: while you are tweaking this - any ideas how [this one](https://github.com/babel/babel/issues/7258) could be solved? The only thing I've came up with was to track holey-ness on bindings for simple cases & bail out for more complex ones.

Comment (loganfsmyth) [reply packages/babel-traverse/src/scope/index.js]: I don't think the helpers themselves would need to change for that. Looks like the core issue is that we call `.concat` on the original. Maybe we just should call `toConsumableArray` on the original data too before `.concat`?

Comment (Andarist) [reply packages/babel-traverse/src/scope/index.js]: It's indeed unrelated to this. I've only reminded myself about the other issue while looking at this one, so thought I'd ask about it.  

> Maybe we just should call toConsumableArray on the original data too before .concat?

That's ofc a possibility - would be good to optimize it though, as i.e. this `.concat` call gets generated often for arguments rest and we could analyze there that it's safe to call `.concat` in such situation. Going to look into this issue again, hopefully soon, gonna try to figure it out to be both safe & optimized if possible.

**Comment by laurre** [comment]
How does this change play with the plugin proposal-object-rest-spread ? I can't get it to work, and I always get the exception mentioned in this issue

**Comment by laurre** [comment]
This change broke my code that was working, no problem with that, I use create-react-app in the latest version that also brings in the plugin proposal-object-rest-spread, it works when I try it with simple objects.

My use case is a bit different. I was using the spread operator with a variable coming from an import, like this (simplified):


```
import { functions as configFunctions } from "../config";

const tab = [ ...configFunctions ];
```

The idea is to get all the functions declared in various modules in a single array. I get the 'Invalid attempt to spread non-iterable instance' error, but I don't get why, isn't the 'functions' variable a normal object ?

In my case the compiled code looks like this:

`var tab = Object(path_to_my_project_node_modules_babel_runtime_helpers_esm_toConsumableArray__WEBPACK_IMPORTED_MODULE_0__["default"])(_location__WEBPACK_IMPORTED_MODULE_8__["actions"]);`

Maybe the way I am doing this is not the most appropriate, but isn't is supposed to work as it was previously ?

**Comment by laurre** [comment]
I have a file '../config/functions.js', in which at the bottom I have:

```
export default {
  function1,
  function2,
  function3
};
```

**Comment by nicolo-ribaudo** [comment]
You can't spread an object into an array[¹]. You should use `var tab = Array.from(Object.values(configFunctions))`. I am surprised that it worked :thinking: 	

[1] You can try this code in any modern browser:
```js
[ ...{ function1() {}, function2() {}, function3() {} } ]
```

**Comment by laurre** [comment]
It works now, thanks, I think it was working before because Babel was to permissive

**Pull request merged**